unitTests/ui/ResultListTest.ts

```diff
@@ -12,169 +12,186 @@ import { AdvancedComponentSetupOptions } from '../MockEnvironment';
+        const newResultsSpy = jasmine.createSpy('newresultsspy');
+        $$(test.cmp.element).on(ResultListEvents.newResultsDisplayed, newResultsSpy);
+        await simulateFetchMore();
+        expect(newResultsSpy).toHaveBeenCalledTimes(1);
+        done();
+      });
 
-        it('should trigger a single new results displayed event when fetching more results', done => {
-          const newResultsSpy = jasmine.createSpy('newresultsspy');
-          $$(test.cmp.element).on(ResultListEvents.newResultsDisplayed, newResultsSpy);
+      it('should log an analytics event when more results are returned', async done => {
+        await simulateFetchMore();
+        expect(test.env.usageAnalytics.logCustomEvent).toHaveBeenCalledWith(
+          analyticsActionCauseList.pagerScrolling,
+          jasmine.any(Object),
+          test.cmp.element
+        );
+        done();
+      });
 
-          Defer.defer(() => {
-            test.cmp.displayMoreResults(10).then(() => {
-              expect(newResultsSpy).toHaveBeenCalledTimes(1);
-              done();
-            });
-          });
+      describe('when infinite scrolling is enabled', () => {
+        beforeEach(async done => {
+          test = Mock.basicComponentSetup<ResultList>(ResultList, { enableInfiniteScroll: true, infiniteScrollPageSize: 1 });
+          mockQueryControllerSearches();
+          await simulateSearch();
+          spyOn(test.cmp as any, 'isScrollingOfResultListAlmostAtTheBottom').and.returnValue(true);
+          done();
         });
 
-        it('should log an analytics event when more results are returned', done => {
-          Defer.defer(() => {
-            test.cmp.displayMoreResults(10).then(() => {
-              expect(test.env.usageAnalytics.logCustomEvent).toHaveBeenCalledWith(
-                analyticsActionCauseList.pagerScrolling,
-                jasmine.any(Object),
-                test.cmp.element
-              );
-              done();
-            });
-          });
+        it('should queue up another scroll when it receives results to fill up the container', async done => {
+          await simulateFetchMore();
+          expect(test.env.queryController.fetchMore).toHaveBeenCalled();
+          done();
         });
 
-        it('should queue up another scroll when it receives results to fill up the container, if infinite scrolling is enabled', done => {
-          test.cmp.options.enableInfiniteScroll = true;
-          Defer.defer(() => {
-            test.cmp.displayMoreResults(10);
-            promiseResults.then(() => {
-              setTimeout(() => {
-                expect(test.env.queryController.fetchMore).toHaveBeenCalled();
-                done();
-              }, 1000);
-            });
-          });
-        });
+        it('should not queue up infinite amount of request if it is trying to fill up the scrolling container', async done => {
+          async function waitForSuccessAndAllAdditionalQueries() {
```

Comment:
This test was waiting one full second for all events to be resolved. I noticed that changing the time it would wait so that it would wait 500ms instead would cause it to detect a smaller number of fetchMore calls.

Since this test already required changes, I decided to change it to reduce the risk of a race condition failing the test.

Reply:
Maybe I'm not understand the code correctly but, what it does right now:

* Execute fetch more 5 times, then resolve the `waitForSuccessAndAllAdditionalQueries`
* Verify that fetch more was called 5 times + once initially.
* But it would not test that the result list would continue calling fetchMore one thousand more time after that verification

So it's kind of a useless test 🤔  ? Or am I misreading the change ?

The intention behind this was to verify that the result list will not go haywire and request results infinitely on its own as fast as it can.

Reply:
The new method detects whether a new fetchMore call was queued, and if so, waits for it to resolves, then verifies if another was queued, and so on infinitely.

If there was no limit, this test would timeout (or never finish, I haven't tested). If there's a limit, this test only passes if the limit is exactly 5.

Prior to my changes, the test was waiting a fixed amount of time which needed to be adjusted to make sure we catch all the calls. There's a chance it would pass even if the actual limit was greater than 5.